### ISSUE TITLE
React to input changes: signal effects vs ngOnChange 

### ISSUE THREAD

**Issue by mauriziocescon** [angular/angular#55027]
### Which @angular/* package(s) are relevant/related to the feature request?

core

### Description

Often in angular, developers write code in order to react to input changes in components. In general, one of the common places used to achieve this goal is the life cycle hook ngOnChange. 

Typically, something like: 
``` typescript
ngOnChanges(changes: SimpleChanges): void {

   // a preliminary "if" lets you target only the input named value
    if (changes['value']) {
        // whenever value changes, do something
    }
}
```
Now, it seems to me the way to achieve the same goal with signals is using effects. So something like: 
``` typescript
valueEffect = effect(() => {

    // the initial assignment makes sure the effect tracks value changes,
    //
    // even if test might not be needed in the untracked func (if you call 
    // a component method using directly this.value()).
    // 
    // Using untracked makes sure nothing else will be tracked except for value. 
    // Moreover, I can set other signals within untracked. 
    const test = this.value();

    untracked(() => {
       // whenever value changes, do something
    });
  });
```

Assuming the considerations above are correct (otherwise sorry for having missed some important points), I have the feeling the signal alternative is a bit strange. Say: effects have a wider scope compared to ngOnChange, but less intuitive to use for a very common thing like "react to an input change". 

I'd personally like to have a more clean approach where (possibly) I clearly state that "an effect should only track this specific signal(s) changes and react only to that". 

Recently, I've noticed there've been some tickets around the same topic: #54985 and #54372. 

Here is a not trivial example to illustrate the differences: https://stackblitz.com/edit/stackblitz-starters-ey1c9p

### Proposed solution

I'd personally like to have a more clean approach where (possibly) I clearly state that "an effect should only track one(many) specific signal(s) changes and react only to that". 

### Alternatives considered

It's mostly a matter of making the intention "react to a specific input change" more clear... no other thing. And again: if I missed some important points, sorry!

**Comment by JeanMeche** [comment]
Hi, about explicit dependencies we had a previous discussion in #50916

**Comment by Harpush** [comment]
@mauriziocescon you can also take my implementation from here: https://github.com/angular/angular/pull/54614#issuecomment-1966283955

**Comment by pkozlowski-opensource** [comment]
So there are multiple topics being discussed here, ranging from working with inputs all the way to the automatic dependencies tracking with signals.

Let me start with re-iterating our position from https://github.com/angular/angular/issues/50916 - we've chosen signals partly _because_ of their automatic tracking nature: we don't want developers to _remember_ and deal with the friction of manual dependency tracking. Having said this people are free to use coding patterns that make dependencies more explicit if this help their code organization.

Now, for the input as signals I would say that reactive primitives (computed, effect) are the way to go. Most of the time we expect people to reach out for a computed and only use effect for truly side-effectful operations. In this sense your example could be re-written as follows:

```typescript
valueEffect = effect(() => {
       // whenever value changes, do something
       doSomething(this.value());
  });
```

I don't think `untracked` is needed by default since:
* using it as a way to allow signal writes is missing the point of the safety check of writing to signals in effects;
* mentioned bugs (ex.: https://github.com/angular/angular/issues/54548) are well, bugs that we've fixed already so those shouldn't factor in the decision;
* `doSomething` might _want_ to respond reactively to the value changes.

I'm happy to discuss more around more concrete examples - here we do have just a short code snippet and this makes it impossible to reason about the general logic

**Comment by mauriziocescon** [comment]
Hey!

Thanks for the reply: much appreciated. Well, I provided a stackblitz at the beginning, but it's just an easy example that can be refactored the way you proposed. 

Considering the inputs part clarified, let me try to make the automatic deps tracking more clear: the logic and functionalities currently available for effects covers all the needs. No disagreement on that. So there isn't anything logically unsupported. 

The problem I see is not about the logic, but what developers will use the most on a daily basis and how effects behave by default. IMHO effects will be massive used (not just rarely), but they don't fulfil (out of the box) common developer needs. 

```typescript
// `doSomething` could do many complex things like calling a method on a global service that
//
// 1) changes a signal as its internal implementation
//  (rxjs optional, signals replacing subjects), 
// 
// 2) that reads some service internal signals to perform some operations 
// (unwanted deps added to valueEffect by mistake)
// 
// 3) that is not under my control in terms of 
// implementation / refactoring (huge enterprise app with ~100 devs)



// default (automatic deps tracking, safe guard on signal writes): I'd say rarely needed 
// and able to track unwanted deps coming from doSomething
valueEffect = effect(() => doSomething(this.value()));


// variation (safe guard on the tracked dep, can write on signals): highly desirable cause 
// I don't need to know anything about the doSomething internal implementation and 
// there's no risks at the level of unwanted deps coming from it
valueEffect = effect(() => {
       this.value();
       untracked(() => doSomething(this.value())
  });
```

I personally believe that the watch func suggestion is what is missing right now (no effect replacement)
https://github.com/angular/angular/issues/54548#issuecomment-1961712985


Here is a stackblitz with the service example mentioned above and lots of comments https://stackblitz.com/edit/stackblitz-starters-csmdrr

Again: the code could be refactored and problems solved. But don't forget about the real life: 
- lots of developers, 
- big app, 
- complexity, 
- calling global services without knowing the internal implementation, 
- junior devs, 
- the quick fix suggestion coming from Error: NG0600 :-) .

In other words: in a world of signals (services / components / ... ) is it really rare writing to signals in effects while reacting (for example) to component input changes? Is the default effect usage the safest in terms of unwanted tracked deps? 

Just my 2 cents!

PS: thanks a lot for the amazing work you're doing!

**Comment by imaksp** [comment]
Hi all, want to add one more common use case, what if I need to call an api function based on some signal field from common store, & that api function will write to **unrelated** signals, here what could be best option untracked or effect with allowSignalWrites ? (I assume untracked will be best option). 

so I think even if untracked usage is discouraged I think many people will need to use it, correct me if I am wrong.

**Comment by mauriziocescon** [comment]
Some recent news somehow related to the topic: https://github.com/proposal-signals/proposal-signals 

Still think a watch (framework) func might be handy and avoid confusion... maybe with some rework on the NG0600 error message suggestions!

**Comment by pkozlowski-opensource** [comment]
Reading more through this issue it doesn't really seem to be about `ngOnChange` but really a more broad discussion about automatic dependencies tracking in effect. As such, let's centralize the discussion in https://github.com/angular/angular/issues/56155.

Duplicate of https://github.com/angular/angular/issues/56155

**Comment by mauriziocescon** [comment]
Yep: basically same request!
